site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+	// match with the criteria above, then the GRPCRoute is not accepted. The
+	// implementation must raise an 'Accepted' Condition with a status of
+	// `False` in the corresponding RouteParentStatus.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Hostnames []Hostname `json:"hostnames,omitempty"`
+
+	// Rules are a list of GRPC matchers, filters and actions.
+        // 
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:default={{matches: {{method: {type: "Exact"}}}}}
+	Rules []GRPCRouteRule `json:"rules,omitempty"`
+}
+
+// GRPCRouteRule defines semantics for matching an gRPC request based on
+// conditions (matches), processing it (filters), and forwarding the request to
+// an API object (backendRefs).
+type GRPCRouteRule struct {
+	// Matches define conditions used for matching the rule against incoming
+	// gRPC requests. Each match is independent, i.e. this rule will be matched
+	// if **any** one of the matches is satisfied.
+	//
+	// For example, take the following matches configuration:
+	//
+	// ```
+	// matches:
+	// - method:
+	//     method: foo.bar
+	//   headers:
+	//     values:
+	//       version: 2
+	// - method:
+	//     service: foo.bar.v2
+	// ```
+	//
+	// For a request to match against this rule, a request should satisfy
+	// EITHER of the two conditions:
+	//
+	// - service of foo.bar AND contains the header `version: 2`
+	// - service of foo.bar.v2
+	//
+	// See the documentation for GRPCRouteMatch on how to specify multiple
+	// match conditions that should be ANDed together.
+	//
+	// If no matches are specified, the matches every gRPC request.
+	//
+	// Proxy or Load Balancer routing configuration generated from GRPCRoutes or HTTPRoutes
+	// MUST prioritize rules based on the following criteria, continuing on
+	// ties. When comparing GRPCRoutes and HTTPRoutes, GRPCRoute matches are first converted
+	// to HTTPRoute matches according to the following scheme:
+	//
+	// * HeaderMatches are translated exactly, since the field are
+	//    one-to-one
+	// * If `MethodMatch.type` is `Exact` and neither the
+	//    `MethodMatch.service` nor `MethodMatch.method` fields are present,
+	//    translate to a `PathMatch` of `{"type": "PathPrefix", value: "/"}`
+	// * If `MethodMatch.type` is `Exact` and both the `MethodMatch.service`
+	//    and `MethodMatch.method` fields are present, translate to a `PathMatch`
+	//    of `{"type": "Exact", value: "/" + service + "/" + method}`
+	// * If `MethodMatch.type` is `Exact` and the `MethodMatch.service`
+	//    field is present but not the `MethodMatch.method` field, translate to a
+	//    `PathMatch` of `{"type": "PathPrefix", value: "/" + service}`
+	// - Otherwise, translate the service and the method individually into
+	//    regular expressions (with an absent field interpreted as `.*`) and
+	//    translate to a `PathMatch` of
+	//    `{"type": "RegularExpression", value: "/": serviceRegex + "/" + methodRegex}`
```

Comment:
To clarify, this would essentially translate an empty Method match with a `foo` Service to `type: PathPrefix, value: /foo`, and a regex would only be required if the match specified a path but not a Service? Not very familiar with gRPC, but is that last option a very common one? Would it be valid to just not support that?

Reply:
> To clarify, this would essentially translate an empty Method match with a foo Service to type: PathPrefix, value: /foo

Correct.

> a regex would only be required if the match specified a path but not a Service

You can imagine many different use cases for regexes. Many gRPC APIs are resource oriented. You may want to route only methods that start with `Get`. In this case, you'd want to do something like:

```
method:
 type: RegularExpression
 service: foo.org
 method: Get.+
```

Or you might do something similar with services. Say you want to route both v1 and v2 of a service. What's nice here is that exact matches are a subset of regular expressions, so the fact that there's only one type for two values here doesn't actually get in the way of expressing the routing logic you want.

> Would it be valid to just not support that?

It would be valid, but it would also be limiting. This is why I have listed "Custom" support for the RegularExpression matcher type.